torch/distributed/_tensor/op_schema.py

```diff
@@ -116,48 +130,21 @@ class OpSchema:
     preserved). It is mainly used by the dispatching logic below to run things like
     sharding propagation.
 
-    Sharding propagation rules registered could utilize this data class and
-    do inplace update some fields (when necessary, i.e shape related ops) to make
-    sure the args/kwargs are legit before passing to the local tensor operator.
-    This is the main reason that we don't freeze this dataclass.
-
-    NOTE: greater access to the operator inputs comes with greater responsibility.
-    Here are some basic rules about what can be used and what can be changed.
+    NOTE: this should be used as a read only data class
+    TODO: make this a frozen dataclass
 
     Args:
-        func_schema: the function schema of the operator
+        op: the operator overload we are intercepting
         args_schema: contains args except that the DTensor args have been replaced
             with its DTensorSpec
         kwargs_schema: contains kwargs except that the DTensor kwargs have been replaced
             with its DTensorSpec
-
-    What can be used:
-        - every attribute within this class could be read to conduct
-          sharding propagation.
-    What can be changed:
-        - only the args_schema and kwargs_schema could be changed.
-        - every non-tensor args could be changed to accomodate for local tensor
-          operations (i.e. for ops like view/reshape/...)
-        - every "DTensorSpec" attribute inside `args_schema`, `kwargs_schema` and
-          `args_spec` SHOULD NOT be updated! DTensorSpec are read only and sharding
-          propagation shouldn't inplace update them, otherwise the input DTensor
-          placements will get implicitly changed and it's error-prone.
     """
 
-    func_schema: torch._C.FunctionSchema
+    op: torch._ops.OpOverload
     args_schema: ArgsType
     kwargs_schema: KwargsType
 
-    is_inplace: bool = False
-    is_out_variant: bool = False
-
-    def __post_init__(self) -> None:
-        # simple analysis of function schema to determine
-        # if this is an inplace/out variant, it might not
-        # be entirely correct, but it's good enough for now.
-        self.is_inplace = self.func_schema.name[-1] == "_"
-        self.is_out_variant = "out" in self.func_schema.overload_name
```

Comment:
Isn't this making the check being called every time in every DTensor Op dispatch? Or we cannot cache the value of these two, so using `__post_init__` does not help that much?

Reply:
This post init just try to analyze the operator we are dispatching over is a inplace op or out variant, but we don't really need this information unless for some certain ops (i.e. when we are checking if the op is inplace or out in pointwise ops), so everytime we form a OpSchema we analyze this become redundant, that's why I am deleting it and put this logic inside the op rules directly.

The caching would cache the op_schema's hash only, and because we don't `hash` these two fields anyways, the cache behavior does not change.